### DISCUSSION TITLE
Why did you choose to go with UTF-16 from the beginning?

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#147] [General]
Let me guess, influenced by Java? Or the Windows API? Qt is the only library that uses UTF-16 for Unicode strings as I know.

**Comment by objeck** [comment]
Funny, I did not. 

The language was originally developed using 1-byte ``char`` values for strings. It was an effort across the codebase and safely convert in a platform-independent way between ``char`` and  ``wchar_t`` strings. 

As you know, ``wchar_t`` is platform dependent; 4-bytes POSIX and 2-bytes Windows.

**Reply by ghost** [reply]
Why didn't you choose UTF-8? Even Windows embraced UTF-8:

https://learn.microsoft.com/vi-vn/windows/apps/design/globalizing/use-utf8-code-page

**Comment by objeck** [comment]
The platform reads and writes UTF-8. 

Once imported, strings are translated into the native ``wchar_t`` format for POSIX UTF-32 and Windows UTF-16., When strings or characters are written, they are converted back to UTF-8.

**Reply by ghost** [reply]
Converting back and forth between encodings will affect performance. Specially when you have to interface with third party libraries (most if not all use UTF-8), only Qt used UTF-16 for Unicode as I know, others are moving away from UTF-16 to embrace UTF-8.

**Comment by objeck** [comment]
A small performance impact given the cross-platform benefits. 

The conversion from UTF-8 to the OS representation (UTF-16 or UTF-32) is a single call using [OS APIs](https://github.com/objeck/objeck-lang/blob/master/core/shared/sys.h#L124). Internally all operations are done natively; no conversions from UTF-X to UTF-Y are required. When strings need to be written, they are converted to UTF-8 or ASCII, the same process the native operating systems follow.

**Comment by objeck** [comment]
Said another way, I choose not to work against the OS, Linux, macOS, and Windows. Instead, embrace their internal representation of character strings and gain performance benefits. All the platform targets support Unicode, a higher-level need.

**Comment by ghost** [comment]
Unices (including Linux and macOS) use UTF-8 for Unicode by default. Windows started from Windows 10 supports UTF-8 for Unicode (they don't introduce new APIs but reuse the old APIs for ANSI, e.g: CreateFileA now will use UTF-8 but not ANSI, after all they all use `char` as the foundation).

Update: Windows doesn't use UTF-8 for Unicode by default, it will need to be specified on the manifest file:

https://learn.microsoft.com/vi-vn/windows/apps/design/globalizing/use-utf8-code-page

p/s: this is only for the sake of updated information, I didn't persuade you to switch to UTF-8.
